### PR TITLE
OAK-10069 : Best practices on how to setup access control external identities

### PR DESCRIPTION
**Pull request by anchela**
hi @joerghoh , i would appreciate if you had time to take a look at this best-practice doc extension for authentication with an external Idp. please let me know if something is missing or not clear.

### CODE REVIEW

File: oak-doc/src/site/markdown/security/authentication/external/bestpractices.md [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+Additional reading: https://docs.oracle.com/en/java/javase/11/security/appendix-b-jaas-login-configuration-file.html#GUID-7EB80FA5-3C16-4016-AED6-0FC619F86F8E
+
+#### Combination with Token Authentication
+
+Whenever JCR sessions created with Oak are short-lived (e.g. only lasting for a single HTTP request) authentication 
+against an external IDP may not perform well. It is therefore recommended to use external authentication in combination 
+with an additional authentication mechanism like e.g. the built-in [token login](../tokenmanagement.html).
+
+Make sure the token-login is 'sufficient' and is evaluated prior to the external login that connects to the external IDP.
```
Comment (joerghoh) [inline oak-doc/src/site/markdown/security/authentication/external/bestpractices.md]: is 'sufficient': it's a bit unclear what it means, can you be a bit more explicit? 

```Make sure that the token-login module is marked with the policy 'sufficient' ```

(I don't know the correct  term for that.)

File: oak-doc/src/site/markdown/security/authentication/external/bestpractices.md [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+   limitations under the License.
+-->
+
+# Best Practices for External Authentication
+
+<!-- MACRO{toc} -->
+
+## Before you get started
+
+Before you get started make sure you are familiar with the basic concepts of JCR authentication, and it's implementation
```
Comment (joerghoh) [inline oak-doc/src/site/markdown/security/authentication/external/bestpractices.md]: Can you please remove the newlines in the paragraphs? A single paragraph followed with a newline (and an editor with soft-wrap support) should be sufficient :)

File: oak-doc/src/site/markdown/security/authentication/external/bestpractices.md [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+Whenever JCR sessions created with Oak are short-lived (e.g. only lasting for a single HTTP request) authentication 
+against an external IDP may not perform well. It is therefore recommended to use external authentication in combination 
+with an additional authentication mechanism like e.g. the built-in [token login](../tokenmanagement.html).
+
+Make sure the token-login is 'sufficient' and is evaluated prior to the external login that connects to the external IDP.
+
+#### Combination with Default Authentication
+
+Oak comes with a default login for user accounts stored and managed inside the JCR content repository. This also includes 
+support for default users like an anonymous guest, and an administrator with full access to the repository. If this is
```
Comment (joerghoh) [inline oak-doc/src/site/markdown/security/authentication/external/bestpractices.md]:
```suggestion
support for default users like 'anonymous' and 'admin' with full access to the repository. If this is 
```

If these are default users coming with Oak, I would name them like that here as well.

File: oak-doc/src/site/markdown/security/authentication/external/bestpractices.md [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+Make sure the token-login is 'sufficient' and is evaluated prior to the external login that connects to the external IDP.
+
+#### Combination with Default Authentication
+
+Oak comes with a default login for user accounts stored and managed inside the JCR content repository. This also includes 
+support for default users like an anonymous guest, and an administrator with full access to the repository. If this is 
+desired, it is recommend to also add the [default `LoginModule`](../default.html#uid_pw) to the JAAS configuration.
+
+The optional order depends on the frequency of default vs external login.
```
Comment (joerghoh) [inline oak-doc/src/site/markdown/security/authentication/external/bestpractices.md]: ```Optimal``` order?

File: oak-doc/src/site/markdown/security/authentication/external/bestpractices.md [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+##### Example JAAS Configuration
+
+The following JAAS configuration can e.g. be used when running an Oak repository with external authentication
```
Comment (joerghoh) [inline oak-doc/src/site/markdown/security/authentication/external/bestpractices.md]:
```suggestion
The following JAAS configuration is an example when running an Oak repository with external authentication 
```

File: oak-doc/src/site/markdown/security/authentication/external/bestpractices.md [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+The following JAAS configuration can e.g. be used when running an Oak repository with external authentication 
+in combination with Apache Sling (see also other [examples](externallogin_examples.html)):
+
+| Ranking | Control Flag | LoginModule Class Name |
+|---------|--------------|------------------------|
+| 300     | OPTIONAL     | org.apache.jackrabbit.oak.spi.security.authentication.GuestLoginModule |
+| 200     | SUFFICIENT   | org.apache.jackrabbit.oak.security.authentication.token.TokenLoginModule | 
+| 150     | SUFFICIENT   | org.apache.jackrabbit.oak.spi.security.authentication.external.impl.ExternalLoginModuleFactory |
+| 100     | SUFFICIENT   | org.apache.jackrabbit.oak.security.authentication.user.LoginModuleImpl |
```
Comment (joerghoh) [inline oak-doc/src/site/markdown/security/authentication/external/bestpractices.md:57]: Can you explain what it does and why it is done this way? That would make that example more useful.

File: oak-doc/src/site/markdown/security/authentication/external/bestpractices.md [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+The following best practices should be followed:
+
+#### External Identity Provider as SSOT
+
+Your external identity provider should be considered the single source of truth (SSOT) for all users and groups defined 
+and managed by it. 
+
+In contrast, the data synchronized into the repository should be considered a volatile cache and ideally 
+are immutable (i.e. only maintained by system sessions in charge of the synchronization).
```
Comment (joerghoh) [inline oak-doc/src/site/markdown/security/authentication/external/bestpractices.md]: "the data ```is``` immutable"

(I am not a native speaker, though....)

File: oak-doc/src/site/markdown/security/authentication/external/bestpractices.md [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+
+#### Access control setup
+
+Synchronized external identities (both in default and in dynamic sync mode) are exposed as principals through the 
+[Principal Management API](../../principal.html) and can be used for access control setup as described 
+in [Access Control Management](../../accesscontrol.html).
+
+See also section [Best Practices for Authorization](../../authorization/bestpractices.html)) for recommendations.
+
+The following options exist if access control setup for external groups should be ready before the first user login:
```
Comment (joerghoh) [inline oak-doc/src/site/markdown/security/authentication/external/bestpractices.md]: the statement "before the first user login" is confusing in 2 ways:
* before that you always mention "user sync", so is the "user login" also a kind of sync?
* "before the first user (ever) logs into that system" or "before the user logs into the system for the first time"?

Can you clarify this?

File: oak-doc/src/site/markdown/security/authentication/external/bestpractices.md [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+
+- Pre-sync external groups to make them available to the principal manager (see next section)
+- Configure [ImportMode](../../accesscontrol/default.html#configuration)=`besteffort` with the default Oak 
+  authorization setup and define access control content for principals before they exist.
+  
+#### Pre-sync of external groups
+
+The following 2 options exist to populate the repository with external group principals before the first user login in
```
Comment (joerghoh) [inline oak-doc/src/site/markdown/security/authentication/external/bestpractices.md]: see above for the term "user login"

File: oak-doc/src/site/markdown/security/authentication/external/bestpractices.md [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+of the regular repository login. See [JMX Synchronization Tool](../usersync.html#jmx-synchronization-tool) and
+[SynchronizationMBean](https://jackrabbit.apache.org/oak/docs/apidocs/org/apache/jackrabbit/oak/spi/security/authentication/external/impl/jmx/SynchronizationMBean.html) 
+for details. This requires the `ExternalIdentityProvider` to implement the methods required to retrieve external 
+identities outside of the authentication step. In general this is the recommended way to pre-sync groups.
+- In case the `ExternalIdentityProvider` does not support user and group sync outside of the regular 
+repository login, external identities can be created using Jackrabbit User Management API. It is important to note though
+that the `rep:externalId` cannot be added or changed once the group has been persisted as the property is system maintained 
+and protected. Note that mistakes in defining the `rep:externalId`, `rep:authorizableId` or `rep:principalName` properties 
+will result in mismatches during authentication, sync and permission evaluation that cannot be corrected.
```
Comment (joerghoh) [inline oak-doc/src/site/markdown/security/authentication/external/bestpractices.md]: What's the recommended way to deal with such situations? Remove the group and recreate it with the correct ```rep:principalId```?

**Review by joerghoh — CHANGES_REQUESTED** [review]

**Comment by anchela** [comment]
@joerghoh , hope to have addressed all your findings. would you have time to check again?

File: oak-doc/src/site/markdown/security/authentication/external/bestpractices.md [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+   limitations under the License.
+-->
+
+# Best Practices for External Authentication
+
+<!-- MACRO{toc} -->
+
+## Before you get started
+
+Before you get started make sure you are familiar with the basic concepts of JCR authentication, and it's implementation in Apache Jackrabbit Oak.
```
Comment (joerghoh) [inline oak-doc/src/site/markdown/security/authentication/external/bestpractices.md]:
```suggestion
Before you get started make sure you are familiar with the basic concepts of JCR authentication, and its implementation in Apache Jackrabbit Oak.
```

**Review by joerghoh — APPROVED** [review]

**Pull request merged**